src/System.Drawing.Common/src/System/Drawing/Icon.Unix.cs

```diff
@@ -707,181 +711,181 @@ private void InitFromStreamWithSize(Stream stream, int width, int height)
+            // Read the icon header
+            using (BinaryReader reader = new BinaryReader(stream))
             {
-                IconDirEntry ide;
-                ide.width = reader.ReadByte();
-                ide.height = reader.ReadByte();
-                ide.colorCount = reader.ReadByte();
-                ide.reserved = reader.ReadByte();
-                ide.planes = reader.ReadUInt16();
-                ide.bitCount = reader.ReadUInt16();
-                ide.bytesInRes = reader.ReadUInt32();
-                ide.imageOffset = reader.ReadUInt32();
-
-                // Vista 256x256 icons points directly to a PNG bitmap
-                // 256x256 icons are decoded as 0x0 (width and height are encoded as BYTE)
-                // and we ignore them just like MS does (at least up to fx 2.0) 
-                // Added: storing data so it can be saved back
-                if ((ide.width == 0) && (ide.height == 0))
-                    ide.ignore = true;
-                else
-                    ide.ignore = false;
-
-                iconDir.idEntries[i] = ide;
-
-                //is this is the best fit??
-                if (!sizeObtained)
+                iconDir.idReserved = reader.ReadUInt16();
+                if (iconDir.idReserved != 0) //must be 0
+                    throw new ArgumentException(SR.Format(SR.InvalidPictureType, "picture", nameof(Icon)));
+
+                iconDir.idType = reader.ReadUInt16();
+                if (iconDir.idType != 1) //must be 1
+                    throw new ArgumentException(SR.Format(SR.InvalidPictureType, "picture", nameof(Icon)));
+
+                dirEntryCount = reader.ReadUInt16();
+                imageData = new ImageData[dirEntryCount];
+                iconDir.idCount = dirEntryCount;
+                iconDir.idEntries = new IconDirEntry[dirEntryCount];
+                // Now read in the IconDirEntry structures
+                for (int i = 0; i < dirEntryCount; i++)
                 {
-                    if (((ide.height == height) || (ide.width == width)) && !ide.ignore)
+                    IconDirEntry ide = new IconDirEntry
+                    {
+                        width = reader.ReadByte(),
+                        height = reader.ReadByte(),
+                        colorCount = reader.ReadByte(),
+                        reserved = reader.ReadByte(),
+                        planes = reader.ReadUInt16(),
+                        bitCount = reader.ReadUInt16(),
+                        bytesInRes = reader.ReadUInt32(),
+                        imageOffset = reader.ReadUInt32()
+                    };
+
+                    // Vista 256x256 icons points directly to a PNG bitmap
+                    // 256x256 icons are decoded as 0x0 (width and height are encoded as BYTE)
+                    // We mark them as ignore and later on we just store the raw bytes to be able to save to a file.
+                    if ((ide.width == 0) && (ide.height == 0))
```

Comment:
Not your code but if you felt like it you could write it as we would, ie
`ide.ignore = (ide.width == 0) && (ide.height == 0);`